test/main.tf

```
@@ -0,0 +1,71 @@
+provider "aws" {
+  region  = "us-west-2"
+  profile = "thunderdome"
+}
+
+resource "aws_cloudwatch_log_group" "group" {
+  name_prefix = "test-subscription"
+}
+
+module "cloudwatch_kinesis_firehose_tf" {
+  source           = "github.com/observeinc/terraform-aws-kinesis-firehose" # tflint-ignore: terraform_module_pinned_source
+  name             = "test-subscription-tf"
+  observe_customer = "101"
+  observe_token    = "fake:token"
+}
+
+module "cloudwatch_logs_subscription" {
+  source           = "../"
+  name             = "test-subscription-tf"
+  filter_name      = "test-tf"
+  kinesis_firehose = module.cloudwatch_kinesis_firehose_tf
+}
+
+resource "aws_iam_role" "subscription_filter" {
+  name_prefix        = "test-subscription"
+  assume_role_policy = <<-EOF
+    {
+      "Version": "2012-10-17",
+      "Statement": [
+        {
+          "Action": "sts:AssumeRole",
+          "Principal": {
+            "Service": "logs.amazonaws.com"
+          },
+          "Effect": "Allow",
+          "Sid": ""
+        }
+      ]
+    }
+  EOF
+}
+
+resource "aws_iam_role_policy_attachment" "subscription_filter" {
+  role       = regex(".*role/(?P<role_name>.*)$", aws_iam_role.subscription_filter.arn)["role_name"]
```

```suggestion
  role       = aws_iam_role.subscription_filter.name
```

If you did want to use this resource, you don't need to reverse the role name out of the ARN. At the end of the `Create` func for a resource, providers should call the `Read` method for that resource:

https://github.com/hashicorp/terraform-provider-aws/blob/8dcdbdaad5d4bb9180d215cea8a9b45e71d3c4e0/internal/service/iam/role.go#L246

On `Read`, the `name` will be set:

https://github.com/hashicorp/terraform-provider-aws/blob/8dcdbdaad5d4bb9180d215cea8a9b45e71d3c4e0/internal/service/iam/role.go#L279

When `name_prefix` is set, `name` is computed:

https://github.com/hashicorp/terraform-provider-aws/blob/8dcdbdaad5d4bb9180d215cea8a9b45e71d3c4e0/internal/service/iam/role.go#L123

A property that is optional and computed means that it _can_ be set, but is computed if unset. So you can still reference it just like properties that are _always_ computed, like `arn`.

---

Cool, I'll change this and see if it works

---

ok it seems this way because the code inside of main.tf depends on a variable `iam_role_arn` which is an arn.

---

in the test case we could change that though